### DISCUSSION TITLE
Lego Technic hub connect to Xbox Controller once and never again.

### DISCUSSION THREAD

**Discussion by GenghisHan-op** [pybricks/support#1784] [General]
Hi,
I tried using pybricks to use my Xbox Controller ( series X ) to control Lego Technic Hub. I can connect without problem the very first attemp. Everything is as it should be. But when I turned off the hub ( long press the button till the light goes off ), it can never connect again.
- I tried pressing the sync button to re-sync but no luck, the controller light keeps flashing quick ( waiting for connect ).
- I reset, update the controller firmware
- I restored the Hub to original firmware, and installed pybrick firmware again ( 3 times )
- I didn't change any code since the last successful connect.
But still no luck.
Please help me.
Thanks for reading.

**Comment by laurensvalk** [comment]
Can you share a screenshot of your code, and also a picture of your robot, showing which devices are plugged in?

**Reply by GenghisHan-op** [reply]
Thanks for the reply, I use the code and the lego set exactly the same with this tutorial on youtube:
https://www.youtube.com/watch?v=Tjg68GV41UU&t=680s
And I can connect no problem first try, but after turning off the hub, nothing works again.

**Reply by BertLindeman** [reply]
Hello,

Sorry for the long post.

The program from the link you gave above, does connect to my technichub each time I start the program.
Maybe a hint: I put the controller ON and after that start the program on the hub.
This makes connecting more stable for me. 

Debugging the xboxcontroller on a technichub is difficult because the hub disconnects from the pc
on xboxcontroller connect.
If there is an error in the program, it might look like the xboxcontroller does not connect.

The program below is simplified to only show the xboxcontroller connection.

- The hub light goes to red on start and switches to green as the xboxcontroller connects.

- A button press changes the hub light shortly to blue and goes back to green.
- The A button flip-flops the led-light.

```python

# original program from https://www.tcm0.tk/2024/02/xbox-bluetooth-controller-konnen-nun.html
# pointed to from https://www.youtube.com/watch?v=Tjg68GV41UU&t=680s

#  /BL simplified to prove xboxcontroller connected or not

from pybricks.hubs import ThisHub
from pybricks.pupdevices import Light
from pybricks.parameters import Color, Port, Button
from pybricks.tools import wait
from pybricks.iodevices import XboxController


hub = ThisHub()
light = Light(Port.C)
light.off()

hub.light.on(Color.RED)
controller = XboxController()
hub.light.on(Color.GREEN)
prev_pressed = None
lightBrightness = 0

while True:
    pressed = controller.buttons.pressed()
    if pressed != prev_pressed:
        hub.light.on(Color.BLUE)
        prev_pressed = pressed

    if Button.MENU in pressed:
        hub.system.shutdown()

    if Button.A in pressed:
        if lightBrightness == 0:
            lightBrightness = 100
        else:
            lightBrightness = 0
        # wait for the button to be released
        while Button.A in pressed:    
            pressed = controller.buttons.pressed()
            wait(10)

    light.on(lightBrightness)
    wait(200)
    hub.light.on(Color.GREEN)

```

The original program has this snippet:
```python
    if ("A" in pressed):
```

That should be:
```python
    if (Button.A in pressed):
```
Similar for the MENU button.

Hope this helps.

Bert

**Reply by GenghisHan-op** [reply]
Thanks for your reply,
I tried with your controller test code and it works perfectly, and my tank.py code suddenly works again.
I don't know where is the problem, now it works just fine.
Ah, one more question: I need to re run the code from pybricks the next time I use my tank, is there any way to just turn on the hub and controller without running the code again from pybricks ?

**Reply by BertLindeman** [reply]
Thank you for the report-back.

Nice that it all works.

>  I need to re run the code from pybricks the next time I use my tank, is there any way to just turn on the hub and controller without running the code again from pybricks ?

No need to re-load the program each time.
Load and save the program on the hub by:
- run the program from pybricks 
- stop the program with the stop-button on pybricks or F6
- power-off the hub by a long press on the hub button (untill the light goes off)

The power-off action stores the program on the hub.

Next time you want to run the program:
- switch the xbox controller on
- press the power button on the hub to switch it on (button blinks blue)
- press the power button on the hub to run the program (button shows steady blue)

That is all.

Enjoy pybricks!

Bert

**Reply by laurensvalk** [reply]
Thanks for helping out Bert!

The short version is: if the hub is on, just press the green button to start the last program again. Nothing else should be needed.

> I don't know where is the problem, now it works just fine.

It looks like that program had a small bug, but it's hard to see since the hub disconnects from the computer. Because you can't see error messages and it would appear as if the controller did not connect.

**Reply by GenghisHan-op** [reply]
Thanks for all your help,
I can code and control my tank without any problems now.
This is such a helpful community.
